### DISCUSSION TITLE
[FEATURE REQ] Additional Options for Time-related Rules (schedule)

### DISCUSSION THREAD

**Discussion by mokahless** [autobrr/autobrr#1657] [Ideas]
**Is your feature request related to a problem? Please describe.**
My personal problem with the current filters is if the times all match (ie. max downloads: 1 Per: Week) and you have a lot of filters, and you configure them initially all at once, the result is a mass downloading all at once. For example, as my above example, you would get a mass download once a week, all competing for bandwidth during that short window after initial torrent upload.

**Describe the solution you'd like**
I think the below would solve my personal problem, while also opening up more customization for all users.

I would like to see a new subsection "schedule rules," at the bottom of the General page.

The schedule section would allow selecting "weekly" or "monthly" (I see those as the immediately most useful but others might be able to use "daily" or "yearly") with a second dropdown to select the day of the week or the day of the month. One then might use this to select "weekly" then "Monday." This filter would then only accept torrents for that filter on Mondays, rejecting them otherwise.

This could either be used alongside the "max downloads per" to better schedule them, or it could be used under the condition of no max limits or shorter limits to rotate between filters. For example, a 7-day schedule of hourly downloading could be achieved that would rotate between 7 different filters by setting each filter to a different day and each "max download per" to 1 per hour.

Another example would be one might use the monthly schedule to balance a number of trackers in such a way as to favour one over the other. One might have, say, 8 trackers. They could set up 8 filters, then set a schedule for each and give them each a different set of days. Tracker 1 could be set for the 1st to the 10th, while tracker 2 might be set for only one day of the month, etc.

To prevent misunderstandings, it might be wise to display a warning to users if they attempt to configure the "schedule" without configuring "max downloads per."

Now that I've given it some thought writing this, I know there are better ways to create a schedule. But one way or the other that is what I would like to request: The ability to schedule times and days in addition to just the download limits.

**Describe alternatives you've considered**
The only thing I can think of at the moment is to time creation of the "max downloads per" sections to simulate what I've described but this doesn't allow all the features that the implementation would. It would only be a workaround to my personal problem.

**Comment by zze0s** [comment]
Advanced scheduling has been [on the board](https://github.com/orgs/autobrr/projects/1/views/4?pane=issue&itemId=1230601) for a long time and mainly because it's not a small quick thing to build. 

In the mean time you can script something with bash and do something with CRON to disable and enable certain filters. The APIs are not documented but check your browser network tab what happens when you toggle on or off a filter. Use that call and an api token and you'll get there.

**Comment by SlovakianCanon** [comment]
I think a nice improvement that wouldn't require a big overhaul of the scheduling as whole would be to add a "custom" option to the "Max downloads per" dropdown. One pattern I often would like to be able to achieve is for instance "max 1 download per 6 hours".

Right now I can only do "max 4 downloads per day" to get a similar result, but what happens is that it grabs 4 releases in a short span in the middle of the night and then doesn't do anything for the rest of the day until the next night.

**Comment by akhdanfadh** [comment]
Second this, that will be indeed a nice improvement.

**Comment by royaltongue** [comment]
Would like this as well. Would be much better on my storage to limit it to something like 100GB/10 days rather than 5 randomly sized (within the set limits) files per day

**Comment by Segelfeldt** [comment]
I agree, currently i have one filter setup as one download per hour, but i would prefer it to be on download each second hour. So i would go from 24 possible downloads down to 12.

Setting 12 daily download will make the load balance go wrong since then it will pick the first 12 the first hours and then nothing during the rest of the day.

**Reply by zze0s** [reply]
That is pretty arbitrary, does it really matter if you do one per hour or half of that?

You can achieve this granularity with an external filter that checks if the time is an even or odd hour. Ask chatgpt to build you a shell script that does it.

**Comment by aciulpan** [comment]
I think I also need this! I need to download one F1 race each week and if I set up 1 download per week then it can become a problem (it already has) if for example only 6 days and 23 hours pass before the next uploaded race. This is a problem for me since I need to be quick to download stuff in order to have a chance to upload and build my ratio.

**Reply by ellidi** [reply]
I relate. I almost don't see a point in running autobrr if a week -1 hour chips away at your ratio.

**Comment by SirHardware** [comment]
Maybe it is possible to first implement a "smaller" scheduling at first if the big change is too much at the moment.
I imagine sth like beeing able to choose when a Day/Week/etc schedule starts (e.g. Filter 1 Week starts at Monday, Filter 2 at Tuesday,...), maybe even a combination like 3 per day but no more than 8 per week.

**Comment by andrechalella** [comment]
Currently autobrr only allows: `X downloads per 1 [hour/day/week/month]`

This is not ideal and leads to the overshoot of downloads that many users complain about in this thread.

What is needed is the inverse: `1 download per Y [hours/days/weeks/months]`. This would correctly spread the downloads, plus permits infinite granularity (`Y` must not be restricted to integers).

**Comment by ForceConstant** [comment]
A couple of questions with current impelmentation, when does the week start? 00:00 on monday? Is there anyway to change the time of the start of the week? With the delay or changing time zone?

**Reply by andrechalella** [reply]
I understand it's a rolling week. That is, a hit now will expire exactly seven days (604 800 seconds) from now. Could be wrong, though.

**Reply by zze0s** [reply]
None of them are rolling, all are fixed. Week starts at 00:00 Monday, ends 23:59 Sunday. There is no way to change this at this time.

What is your use case where this would be beneficial?

**Reply by andrechalella** [reply]
Thanks. My pet peeve is the "overshoot" of downloads when the counter is reset, thus my suggestion above ("1 download per Y").

**Comment by ForceConstant** [comment]
For optimal seeding, I would like to trigger during specific peak times of
day/week

**Comment by ElectronicEconomy317** [comment]
This is a good add-on! Can't wait to see it!
